resources/js/charts/changelog-chart.ts

```diff
@@ -0,0 +1,308 @@
+
+    const sortedData = groupBy(rawData, 'created_at');
+
+    const data = map(sortedData, (values, timestamp) => {
+      let sum = sumBy(values, 'user_count');
+
+      if (sum === 0) {
+        let fakedVal = resetLabel != null
+          ? values.find((val) => val.label === resetLabel)
+          : null;
+
+        if (fakedVal == null) {
+          fakedVal = last(values);
+          resetLabel = fakedVal?.label ?? null;
+        }
+
+        if (fakedVal) fakedVal.user_count = 1;
+        sum = 1;
+      } else {
+        hasData ??= true;
+      }
+
+      // parse date stored in strings to JS Date object for use by
+      // d3 domains, and format it into a string shown on the tooltip
+      const m = moment(values[0].created_at);
+
+      const obj: DataObj = {
+        builds: {},
+        created_at: timestamp,
+        date: m.toDate(),
+        date_formatted: m.format('YYYY/MM/DD'),
+      };
+
+      for (const val of values) {
+        val.normalized = val.user_count / sum;
+        obj.builds[String(val.label)] = val;
+      }
+
+      return obj;
+    });
+
+    return { data, hasData };
+  }
+
+  private setDimensions() {
+    const areaDims = this.area.node()?.getBoundingClientRect();
+    if (!areaDims) return;
+
+    this.width = areaDims.width;
+    this.height = areaDims.height;
+  }
+
+  private setHoverAreaSize() {
+    this.hoverArea.attr('width', this.width).attr('height', this.height);
+  }
+
+  private setScalesRange() {
+    this.scales.x.range([0, this.width]).domain([
+      first(this.data[0])?.data.date ?? 0,
+      last(this.data[0])?.data.date ?? 0,
+    ]);
+
+    this.scales.y.range([0, this.height]).domain([0, 1]);
+
+    this.scales.class.range(
+      this.chartConfig.order.map((d, i) =>
+        // rotate over available build ids (0-6) when the amount of builds
+        // exceeds the available amount of colors
+        this.chartConfig.stream_name != null
+          ? `${this.chartConfig.stream_name}-build-${i % 7}`
+          : kebabCase(d ?? ''),
+      ),
+    ).domain(this.chartConfig.order.map(String));
+  }
+
+  private setSvgSize() {
+    this.svg.attr('width', this.width).attr('height', this.height);
+  }
+
+  private readonly tooltipHide = () => fadeOut(this.tooltipContainer.node());
+
+  private readonly tooltipMove = (event: Event) => {
+    const mousePos = d3.pointer(event);
+    this.x = this.scales.x.invert(mousePos[0]);
+    this.y = mousePos[1] / this.height;
+
+    this.tooltipShow();
+
+    clearTimeout(this.autoHideTooltip);
+    this.autoHideTooltip = setTimeout(this.tooltipHide, 3000);
+
+    this.tooltipPosition();
+  };
+
+  private tooltipPosition() {
+    const { x, y } = this;
+
+    if (x == null || y == null) return;
+
+    const pos = d3
+      .bisector<Datum, Date>((d) => d.data.date)
+      .left(this.data[0], x);
+
+    let dataRow = 0;
+    let currentLabel = '';
+    let labelModifier = '';
+
+    for (const [i, el] of this.data.entries()) {
```

Comment:
if you look closer, the `i` isn't needed; just assign the `el[pos]` to something directly.